R/derive_vars_dtm_to_tm.R

```diff
@@ -0,0 +1,64 @@
+#' Derive time variables from datetime variables
+#'
+#' This function creates a time as output from a datetime variable
+#'
+#' @param dataset Input dataset
+#'
+#' @param source_vars A list of datetime variables from which time are to be extracted
+#'
+#' @param hms_flag A `"Y"` scalar character. If `"Y"`, time is represented as `"HH:MM:SS"`
+#'  otherwise they are presented as `"HH:MM"`.
+#'  Optional; default is `NULL`.
+#'
+#' @author Teckla Akinyi
+#'
+#' @return A data frame containing all observations and variables of the input
+#'   dataset and adds the corresponding time variable of all specified datetime variables from
+#'   the source_vars option with the correct name.
+#'
+#' @keywords ADaM Timing Date
+#'
+#' @export
+#'
+#' @examples
+#' library(dplyr, warn.conflicts = FALSE)
+#' library(lubridate, warn.conflicts = FALSE)
+#'
+#' adcm <- tibble::tribble(
+#'  ~STUDYID, ~USUBJID, ~TRTSDTM,               ~ASTDTM,               ~AENDTM,
+#'  "TEST01", "PAT01",  "2012-02-25 23:41:10", "2012-02-28 19:03:00", "2013-02-25 23:32:16",
+#'  "TEST01", "PAT01",  "",                    "2012-02-28 19:00:00", "",
+#'  "TEST01", "PAT01",  "2017-02-25 23:00:02", "2013-02-25 19:00:15", "2014-02-25 19:00:56",
+#'  "TEST01", "PAT01",  "2017-02-25 16:00:00", "2017-02-25 14:25:00", "2017-03-25 23:00:00",
+#'  "TEST01", "PAT01",  "2017-02-25 16:05:17", "2017-02-25 14:20:00", "2018-04-29 14:06:45",
+#'  ) %>% mutate(
+#'  TRTSDTM = lubridate::as_datetime(TRTSDTM),
+#'  ASTDTM = lubridate::as_datetime(ASTDTM),
+#'  AENDTM = lubridate::as_datetime(AENDTM)
+#'  )
+#'
+#' derive_vars_dtm_to_tm(adcm,
+#'                       vars(TRTSDTM, ASTDTM, AENDTM))
+#'
+#' derive_vars_dtm_to_tm(adcm,
+#'                      vars(TRTSDTM, ASTDTM, AENDTM),
+#'                      hms_flag="Y")
+
+derive_vars_dtm_to_tm <- function(dataset,
+                                  source_vars,
+                                  hms_flag=NULL) {
```

Comment:
If we want to keep `hms_flag` then I'd suggest to use `TRUE` and `FALSE` rather than `"Y"` and `NULL`. However, I'm not quite sure about this functionality. I would expect to get the "whole" time component from the datetime rather than only hour and minute. This seems to be more about display (kind of like a SAS format) rather than the actual data. I'd suggest to make use of the `hms` package. Any thought @aehmann-gsk @bundfussr?

Reply:
According to ADaMIG `*TM` variables must be numeric. I.e., we should update the function such that the new variables are numeric and then we no longer need the `hms_flag`.

Reply:
This is a bit of a challenge since for numeric times, they are being displayed as number of seconds since 1970-01-01. I can implement this using the hms package resulting in class "hms" and displaying as HH:MM:SS or the initial function resulting in class "character" with option to display as HH:MM or HH:MM:SS. Otherwise, I do not know that R can handle times akin to SAS where it is numeric but displays as HH:MM. Open to exploring other packages/suggestions

Reply:
The `hms` package looks like what we need. @thomas-neitmann , @aehmann-gsk , do you agree?

Do you now if the `hms` class becomes a time variable when exported as a SAS dataset?

Reply:
Updated to use hms. So far, I have not successfully read in the time vars from the xpt. They do not compute, but will explore more

Reply:
Hi @teckla-gsk , @bundfussr, @esimms999-gsk 

I have make a try with the `haven` package and there seems to be no issue reading the date nor time in XPT/ SAS
```
library(haven) 
library(diffdf)

input <- tibble::tribble(
  ~STUDYID, ~USUBJID, ~TRTSDTM, ~ADTM,               
  "TEST01", "PAT01",  "2012-02-25 23:41:10", "2012-02-28 19:03:00") %>% 
  mutate(TRTSDTM = as_datetime(TRTSDTM), ADTM = as_datetime(ADTM))

actual_output <- derive_vars_dtm_to_tm(input, vars(TRTSDTM, ADTM))

#Convert to XPT
write_xpt(actual_output, "test.xpt",version = 5)           

#Compare data before xpt with data read from xpt
actual_output_fromxpt<-read_xpt("test.xpt")     

diffdf(actual_output, actual_output_fromxpt)
```
The only diff between the 2 dataframes is the `format.sas` attributes, and I guess this will be dealt with once we tackle the metadata stuff.

```
There are columns in BASE and COMPARE with differing attributes !!
All rows are shown in table below

  ================================================
   VARIABLE  ATTR_NAME   VALUES.BASE  VALUES.COMP 
  ------------------------------------------------
     ADTM    format.sas     NULL       DATETIME   
     ATM     format.sas     NULL         TIME     
   TRTSDTM   format.sas     NULL       DATETIME   
    TRTSTM   format.sas     NULL         TIME     
  ------------------------------------------------
```

I was also been able to read the XPT file in SAS with no problem
```
libname xptfile xport 'Path\test.xpt' access=readonly;
proc copy inlib=xptfile outlib=work;
run;
```
![image](https://user-images.githubusercontent.com/57490460/135587683-2c81ec0c-1cbf-4c4c-89f5-3aec9d26981e.png)

So i guess we are good :)

Maybe 1 comment regarding the function itself: i have first tried with 1 DTM variable as input, but when only 1 DTM is specified the new variable is misspelled as **'new'**
![image](https://user-images.githubusercontent.com/57490460/135588937-41d6b3c8-0f3c-40e0-a440-959f0f87fa75.png)

Reply:
@kabis-ops thanks for testing it and the comment. I will update it based on your comment :)